File: SwiftAlgorithm/BOJ/10000/13460.swift
//
//  13460.swift
//  SwiftAlgorithm
//
//  Created by 유정주 on 2023/03/23.
//

import Foundation

struct Item {
    var redX: Int
    var redY: Int
    var blueX: Int
    var blueY: Int
    var count: Int
}

let dx = [-1, 1, 0, 0], dy = [0, 0, -1, 1]

let nm = readLine()!.components(separatedBy: " ").map { Int($0)! }
let n = nm[0], m = nm[1]
let xRange = (0..<n), yRange = (0..<m)

var start: Item = .init(redX: -1, redY: -1, blueX: -1, blueY: -1, count: 0)
var map: [[String]] = []
for i in 0..<n {
    let input = Array(readLine()!).map { String($0) }
    map.append(input)
    
    if start.redX == -1 || start.blueX == -1 {
        for j in 0..<input.count {
            if input[j] == "R" {
                start.redX = i
                start.redY = j
            } else if input[j] == "B" {
                start.blueX = i
                start.blueY = j
            }
        }
    }
}

var queue: [Item] = [start]
var index = 0
var visited: [[[[Bool]]]] = Array(repeating: Array(repeating: Array(repeating: Array(repeating: false, count: m), count: n), count: m), count: n)
visited[start.redX][start.redY][start.blueX][start.blueY] = true

var result = -1
while index < queue.count {
    let node = queue[index]
    index += 1
    
    if node.count > 10 {
        break
    }
    
    if map[node.redX][node.redY] == "O" && map[node.blueX][node.blueY] != "O" {
        result = node.count
        break
    }
    
    for i in 0..<4 {
        var nextRedX = node.redX
        var nextRedY = node.redY
        var nextBlueX = node.blueX
        var nextBlueY = node.blueY
        
        //빨간 구슬 기울기
        while xRange ~= nextRedX, yRange ~= nextRedY {
            if map[nextRedX][nextRedY] != "#" && map[nextRedX][nextRedY] != "O" {
                nextRedX += dx[i]
                nextRedY += dy[i]
            } else if map[nextRedX][nextRedY] == "#" {
                nextRedX -= dx[i]
                nextRedY -= dy[i]
                break
            } else {
                break
            }
        }
        
        //파란 구슬 기울기
        while xRange ~= nextBlueX, yRange ~= nextBlueY {
            if map[nextBlueX][nextBlueY] != "#" && map[nextBlueX][nextBlueY] != "O" {
                nextBlueX += dx[i]
                nextBlueY += dy[i]
            } else if map[nextBlueX][nextBlueY] == "#" {
                nextBlueX -= dx[i]
                nextBlueY -= dy[i]
                break
            } else {
                break
            }
        }
        
        //겹치는 경우
        if nextRedX == nextBlueX && nextRedY == nextBlueY {
            if map[nextRedX][nextRedY] != "O" {
                //더 많이 움직인 구슬을 반대 방향으로 한 칸 이동
                let redDist = abs(nextRedX - node.redX) + abs(nextRedY - node.redY)
                let blueDist = abs(nextBlueX - node.blueX) + abs(nextBlueY - node.blueY)
                
                if redDist > blueDist {
                    nextRedX -= dx[i]
                    nextRedY -= dy[i]
                } else {
                    nextBlueX -= dx[i]
                    nextBlueY -= dy[i]
                }
            }
        }
        
        if !visited[nextRedX][nextRedY][nextBlueX][nextBlueY] {
            visited[nextRedX][nextRedY][nextBlueX][nextBlueY] = true
            let next = Item(redX: nextRedX, redY: nextRedY, blueX: nextBlueX, blueY: nextBlueY, count: node.count + 1)
            queue.append(next)
        }
    }
}

print(result)

File: Python/BOJ/시뮬레이션/13460.py
//
//  13460.py
//  Algorithm
//
//  Created by 유정주 on 2023/06/04.
//

import sys; readLine = sys.stdin.readline
from collections import deque

dx = [-1, 1, 0, 0]; dy = [0, 0, -1, 1]

n, m = map(int, readLine().split())
board = [list(map(str, readLine().rstrip())) for _ in range(n)]

start = [0, 0, 0, 0, 0]
for i in range(n):
  for j in range(m):
    if board[i][j] == "R":
      start[0] = i
      start[1] = j
    elif board[i][j] == "B":
      start[2] = i
      start[3] = j

q = deque([start])
visited = [[[[False for _ in range(m)] for _ in range(n)] for _ in range(m)] for _ in range(n)]
visited[start[0]][start[1]][start[2]][start[3]] = True

result = -1
while q:
  redX, redY, blueX, blueY, count = q.popleft()

  if count > 10:
    break

  if board[redX][redY] == "O" and board[blueX][blueY] != "O":
    result = count
    break

  for i in range(4):
    nxRed = redX
    nyRed = redY
    nxBlue = blueX
    nyBlue = blueY

    while 0 <= nxRed < n and 0 <= nyRed < m:
      if board[nxRed][nyRed] != "#" and board[nxRed][nyRed] != "O":
        nxRed += dx[i]
        nyRed += dy[i]
      elif board[nxRed][nyRed] == "#":
        nxRed -= dx[i]
        nyRed -= dy[i]
        break
      else:
        break

    while 0 <= nxBlue < n and 0 <= nyBlue < m:
      if board[nxBlue][nyBlue] != "#" and board[nxBlue][nyBlue] != "O":
        nxBlue += dx[i]
        nyBlue += dy[i]
      elif board[nxBlue][nyBlue] == "#":
        nxBlue -= dx[i]
        nyBlue -= dy[i]
        break
      else:
        break

    if nxRed == nxBlue and nyRed == nyBlue:
      if board[nxRed][nyRed] != "O":
        redDist = abs(nxRed - redX) + abs(nyRed - redY)
        blueDist = abs(nxBlue - blueX) + abs(nyBlue - blueY)

        if redDist > blueDist:
          nxRed -= dx[i]
          nyRed -= dy[i]
        else:
          nxBlue -= dx[i]
          nyBlue -= dy[i]

    if not visited[nxRed][nyRed][nxBlue][nyBlue]:
      visited[nxRed][nyRed][nxBlue][nyBlue] = True
      q.append([nxRed, nyRed, nxBlue, nyBlue, count + 1])

print(result)
